### PR TITLE
[HttpKernel] Deprecate some compiler passes in favor of tagged iterator args

### PR DESCRIPTION
**Pull request by nicolas-grekas**
| Q             | A
| ------------- | ---
| Branch?       | 3.4
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | yes
| Tests pass?   | yes
| Fixed tickets | -
| License       | MIT
| Doc PR        | -

More code that we can drop :)

### CODE REVIEW

**Comment by nicolas-grekas** [comment]
when merging into master, symfony/dependecy-injection should be removed from Config's composer.json

File: src/Symfony/Component/HttpKernel/CacheClearer/ChainCacheClearer.php [hunk 1]
```diff
@@ -47,9 +47,13 @@ public function clear($cacheDir)
...
      *
      * @param CacheClearerInterface $clearer
+     *
+     * @deprecated since version 3.4, to be removed in 4.0, inject the list of clearers as a constructor argument instead.
      */
     public function add(CacheClearerInterface $clearer)
     {
+        @trigger_error(sprintf('The "%s()" method is deprecated since version 3.4 and will be removed in 4.0, inject the list of clearers as a constructor argument instead.', __METHOD__), E_USER_DEPRECATED);
+
         $this->clearers[] = $clearer;
```
Comment (chalasr) [inline src/Symfony/Component/HttpKernel/CacheClearer/ChainCacheClearer.php]: isn't`$this->clearers` an iterator now?

Comment (nicolas-grekas) [reply src/Symfony/Component/HttpKernel/CacheClearer/ChainCacheClearer.php]: Is it, but here is my reasoning: this method is deprecated, and no code exists yet that uses both the method and an iterator. No new code should be written that would use both an iterator, and the deprecated method. Thus, all is fine :)

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @nicolas-grekas.

**Comment by Tobion** [comment]
Ref PR introducing this feature: #22200

**Pull request merged**